assets/js/googlesitekit/data/create-notifications-store.js

```diff
@@ -0,0 +1,172 @@
+import { keyBy, omit } from 'lodash';
+import { get } from 'googlesitekit-api';
+
+const ADD_NOTIFICATION = 'ADD_NOTIFICATION';
+const REMOVE_NOTIFICATION = 'REMOVE_NOTIFICATION';
+const FETCH_NOTIFICATIONS = 'FETCH_NOTIFICATIONS';
+const RECEIVE_NOTIFICATIONS = 'RECEIVE_NOTIFICATIONS';
+const RECEIVE_NOTIFICATIONS_FAILED = 'RECEIVE_NOTIFICATIONS_FAILED';
+
+// This should remain private for now, hence not be exported on Data.
+export const createNotificationsStore = ( type, identifier, datapoint ) => {
+	const INITIAL_STATE = {
+		serverNotifications: undefined,
+		clientNotifications: {},
+		isFetchingNotifications: false,
+	};
+
+	const actions = {
+		addNotification( notification ) {
+			return {
+				notification,
+				type: ADD_NOTIFICATION,
+			};
+		},
+
+		removeNotification( id ) {
+			return {
+				id,
+				type: REMOVE_NOTIFICATION,
+			};
+		},
+
+		fetchNotifications() {
+			return {
+				payload: {},
+				type: FETCH_NOTIFICATIONS,
+			};
+		},
+
+		receiveNotifications( notifications ) {
+			invariant( notifications, 'notifications is required.' );
+
+			return {
+				payload: { notifications },
+				type: RECEIVE_NOTIFICATIONS,
+			};
+		},
+
+		receiveNotificationsFailed( error ) {
+			invariant( error, 'error is required.' );
+
+			return {
+				payload: {},
+				type: RECEIVE_NOTIFICATIONS_FAILED,
+			};
+		},
+	};
+
+	const controls = {
+		[ FETCH_NOTIFICATIONS ]: () => {
+			return get( type, identifier, datapoint );
+		},
+	};
+
+	const reducer = ( state = INITIAL_STATE, action ) => {
+		switch ( action.type ) {
+			case ADD_NOTIFICATION: {
+				const { notification } = action;
+
+				return {
+					...state,
+					clientNotifications: {
+						...state.clientNotifications,
+						...keyBy( [ notification ], 'id' ),
+					},
+				};
+			}
+
+			case REMOVE_NOTIFICATION: {
+				const { id } = action;
+
+				return {
+					...state,
+					clientNotifications: omit( state.clientNotifications, [ id ] ),
+				};
+			}
+
+			case FETCH_NOTIFICATIONS: {
+				return {
+					...state,
+					isFetchingNotifications: true,
+				};
+			}
+
+			case RECEIVE_NOTIFICATIONS: {
+				const { notifications } = action.payload;
+
+				return {
+					...state,
+					isFetchingNotifications: false,
+					serverNotifications: {
+						...keyBy( notifications, 'id' ),
+					},
+				};
+			}
+
+			case RECEIVE_NOTIFICATIONS_FAILED: {
+				return {
+					...state,
+					isFetchingNotifications: false,
+				};
+			}
+
+			default: {
+				return { ...state };
+			}
+		}
+	};
+
+	const resolvers = {
+		*getNotifications() {
+			try {
+				const notifications = yield actions.fetchNotifications();
+				return actions.receiveNotifications( notifications );
+			} catch ( err ) {
+				return actions.receiveNotificationsFailed( err );
+			}
+		},
+	};
+
+	const selectors = {
+		getNotifications( state ) {
+			const { serverNotifications, clientNotifications } = state;
+
+			return Object.values( {
+				...( serverNotifications || {} ),
```

Comment:
Storing these separately results in notifications appearing, potentially, out-of-order because server notifications always come "first". I'd expect the selector to return notifications in their intended order, so another reason to combine the notifications and mark them with a marker instead.

Reply:
There is no notion of order in our notifications currently, so I think that's perfectly fine for now. Again, this is purely internal, from the outside there must not be a differentiation for now. Except that right now server-side notifications cannot be removed.